libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -700,3 +705,88 @@ AP_GPS::_broadcast_gps_type(const char *type, uint8_t instance, int8_t baud_inde
     }
     GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, buffer);
 }
+
+
+/* 
+   re-assemble GPS_RTCM_DATA message
+ */
+void AP_GPS::handle_gps_rtcm_data(const mavlink_message_t *msg)
+{
+    mavlink_gps_rtcm_data_t packet;
+    mavlink_msg_gps_rtcm_data_decode(msg, &packet);
+
+    if (packet.len > MAVLINK_MSG_GPS_RTCM_DATA_FIELD_DATA_LEN) {
+        // invalid packet
+        return;
+    }
+    
+    if ((packet.flags & 1) == 0) {
+        // it is not fragmented, pass direct
+        inject_data_all(packet.data, packet.len);
+        return;
+    }
+
+    // see if we need to allocate re-assembly buffer
+    if (rtcm_buffer == nullptr) {
+        rtcm_buffer = (struct rtcm_buffer *)calloc(1, sizeof(*rtcm_buffer));
+        if (rtcm_buffer == nullptr) {
+            // nothing to do but discard the data
+            return;
+        }
+    }
+
+    uint8_t fragment = (packet.flags >> 1U) & 0x03;
+    uint8_t sequence = (packet.flags >> 3U) & 0x1F;
+
+    // see if this fragment is consistent with existing fragments
+    if (rtcm_buffer->fragments_received &&
+        (rtcm_buffer->sequence != sequence ||
```

Comment:
> the highest obligation we have at the driver level is to attempt to assemble and transmit the most recent data we have received

I completely agree and that is why I don't like this simple approach. You can receive a single old fragment and you drop an almost full current packet.

I appreciate your analysis of the transport protocols, but the truth is we can't rely on that. My biggest complain with MAVLink is that it doesn't enforce **anything** on the transport level. In my perspective that means we need to address any transport problem in the application layer.

> My biggest concern with trying to limit how fast the sequnce can advance is that depending on what you want to use this with

We would only be limiting when it wraps around - and my proposed number for that is almost 1/3 of the available sequence numbers. We can increase it even more, but I believe that if we lost 10 sequence packets in a short time (otherwise the "timer" would fix it) we will have other problems.

Now, another question: do the non-fragmented packets also use the sequence number?

Sure, let's discuss this in the dev call.